### PR TITLE
[WIP] Another pass at Kotlin Gradle MPP 1.3

### LINKED ISSUE #537: Migrate to the new Kotlin multiplatform Gradle plugin
Here's the guide: https://kotlinlang.org/docs/reference/building-mpp-with-gradle.html
Thread announcing the new plugin: https://discuss.kotlinlang.org/t/kotlin-1-3-m2-new-multiplatform-projects-model/9264
Bunch of examples here (marked obsolete but whevs): https://github.com/h0tk3y/k-new-mpp-samples

Would be great if we could keep the current project structure to avoid polluting Git history and making it hard for in-flight PRs to rebase, but I'm not sure it's possible.

### PR DESCRIPTION
**Pull request by Egorand**
- [x] Figure out a way to run Java tests (https://youtrack.jetbrains.com/issue/KT-28985)
- [x] Configure JS testing
- [x] Native configuration
- [x] Ensure tests run on all platforms (~https://youtrack.jetbrains.com/issue/KT-26547~, ~https://youtrack.jetbrains.com/issue/KT-29039~, https://youtrack.jetbrains.com/issue/KT-29162)
- [x] JVM: Japicmp
- [x] JVM: JMH
- [x] JVM: Animal Sniffer
- [x] JVM: Shadow
- [x] Dokka setup (https://youtrack.jetbrains.com/issue/KT-27901)
- [x] Publishing setup

Closes #537

### CODE REVIEW

**Comment by Egorand** [comment]
Relying on the old Gradle plugin is a major roadblock at this point since I guess all tooling updates will be targeting the new plugin from now on. This feels like a lot of work, mainly because of how different the new plugin is, and how limiting it feels when facing non-standard use-cases. Let's use this PR as the central point for the migration efforts, please contribute by targeting this branch.

**Comment by Egorand** [comment]
Using `presets.jvmWithJava` doesn't seem to run Java tests for some reason..

**Comment by Egorand** [comment]
Interesting, we've got failing Native tests. Reporting seems to be broken in the plugin, so I can only see the tests that failed. Here are the ones that fail on my Mac:
```
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] okio.CommonByteStringTest.ByteString.decodeBase64WithWhitespace
[  FAILED  ] okio.CommonByteStringTest.ByteString.ignoreUnnecessaryPadding
[  FAILED  ] okio.CommonByteStringTest.OkioEncoder.decodeBase64WithWhitespace
[  FAILED  ] okio.CommonByteStringTest.OkioEncoder.ignoreUnnecessaryPadding
```

**Comment by bnorm** [comment]
I believe this is because Kotlin/Native changed ByteString to String UTF-8 translation with 1.3. Changing to use our internal UTF-8 translation for Kotlin/Native like Kotlin/JS in `-Platform.kt` fixes the failing native unit tests.

**Comment by Egorand** [comment]
Shelving this until 1.3.20, looks like the DSL will change.

**Comment by Egorand** [comment]
The publishing setup I ended up with actually publishes artifacts for all supported platforms. I can figure out how to disable this, but I'd be curious to try the current setup against a few multiplatform samples.

**Comment by Egorand** [comment]
I think this is ready for testing! Would love some help in making sure I haven't missed anything and everything is still being built, tested and published properly.

File: okio/build.gradle [hunk 1]
```diff
@@ -1,7 +1,91 @@
...
+    }
+    commonTest {
+      kotlin.srcDir('src/test/kotlin')
+      dependencies {
+        implementation deps.kotlin.test.common
+        implementation deps.kotlin.test.annotations
+      }
+    }
+    jvmMain {
+      kotlin.srcDir('jvm/src/main/java')
```
Comment (bnorm) [inline okio/build.gradle]: Do you also want the resource directory (`jvm/src/main/resources`) for the progruad rules?

File: okio/build.gradle [hunk 2]
```diff
@@ -1,7 +1,91 @@
...
+    compilations.main.outputKinds 'framework'
+  }
+  iosArm64() {
+    compilations.main.outputKinds 'framework'
+  }
+  // Required to generate tests tasks: https://youtrack.jetbrains.com/issue/KT-26547
+  linuxX64()
+  macosX64()
+  mingwX64('winX64')
+  mingwX64()
```
[the remark this answers is not in the document]
Comment (Egorand) [reply okio/build.gradle]: Removed!

File: okio/build.gradle [hunk 3]
```diff
@@ -1,7 +1,91 @@
...
+    commonTest {
+      kotlin.srcDir('src/test/kotlin')
+      dependencies {
+        implementation deps.kotlin.test.common
+        implementation deps.kotlin.test.annotations
+      }
+    }
+    jvmMain {
+      kotlin.srcDir('jvm/src/main/java')
+      dependsOn commonMain
```
[the remark this answers is not in the document]
Comment (Egorand) [reply okio/build.gradle]: I'll remove them

File: okio/build.gradle (hunk 1 above)
Comment (Egorand) [reply okio/build.gradle]: Yep, good call!

**Pull request merged**